### PR TITLE
[Modelzoo] Rewrite DIN model.

### PR DESCRIPTION
**Pull request by Duyi-Wang**
DIEN's code has been rewritten to use latest API and enable distributed training, and BF16 feature is enabled.
Issue:
- acc is very low, only 0.5.
- auc is only 0.7
- BF16 almost no profit.

Gsteps:
positive order, sequence length from small to large.:
35.280950 dp_bf16 (1.241x)
36.179800 dp_fp32 (1.273x)
28.421480 tf (baseline)

ACC/AUC;
tf:
ACC = 0.5001402497291565
AUC = 0.7169774770736694
deeprec fp32:
ACC = 0.5008084774017334
AUC = 0.7026781439781189
deeprec bf16:
ACC = 0.5005197525024414
AUC = 0.7146108746528625

TODO:
- optimize acc & auc

### CODE REVIEW

File: modelzoo/DIN/train.py [hunk 1]
```diff
@@ -1,387 +1,739 @@
...
+                        help='Full path to checkpoints input/output directory',
                         required=False)
+    parser.add_argument('--deep_dropout',
+                        help='Dropout regularization for deep model',
+                        type=float,
+                        default=0.0)
+    parser.add_argument('--learning_rate',
+                        help='Learning rate for model',
+                        type=float,
+                        default=0.001)
```
Comment (aalbersk) [inline modelzoo/DIN/train.py:452]: Why is learning rate so low for DIN model? How does the model behave for higher values of learning rate?

File: modelzoo/DIN/train.py [hunk 2]
```diff
@@ -1,387 +1,739 @@
...
+        self.logits = tf.squeeze(self.logits)
+        loss = tf.losses.sigmoid_cross_entropy(
+            self.label,
+            self.logits,
+            scope='loss',
+            reduction=tf.losses.Reduction.SUM_OVER_BATCH_SIZE)
+        tf.summary.scalar('loss', loss)
+
+        self.global_step = tf.train.get_or_create_global_step()
+        optimizer = tf.train.AdamOptimizer(learning_rate=self.learning_rate)
```
Comment (aalbersk) [inline modelzoo/DIN/train.py:413]: Have you tried running optimizer with BinaryCrossentropy() instead of Adam?

File: modelzoo/DIN/train.py [hunk 3]
```diff
@@ -1,387 +1,739 @@
...
+    no_of_training_examples = sum(1 for line in open(train_file))
+    no_of_test_examples = sum(1 for line in open(test_file))
+    print("Numbers of training dataset is {}".format(no_of_training_examples))
+    print("Numbers of test dataset is {}".format(no_of_test_examples))
+
+    # set params
+    # set batch size & steps
+    batch_size = args.batch_size
+    if args.steps == 0:
+        no_of_epochs = 3
```
Comment (aalbersk) [inline modelzoo/DIN/train.py:512]: Why is number of epochs hard-coded to 3 in the script for args.steps equal 0?

File: modelzoo/DIN/train.py (hunk 1 above)
Comment (Duyi-Wang) [reply modelzoo/DIN/train.py:452]: I got better acc/auc with this learning rate compared to a higher value.

File: modelzoo/DIN/train.py (hunk 2 above)
Comment (Duyi-Wang) [reply modelzoo/DIN/train.py:413]: DeepRec does some optimization on Adam optimizer, so we can see that DeepRec has a more obvious advantage compared to stock tf when using Adam optimizer. In Ali's model, they use adam optimizer.

File: modelzoo/DIN/train.py (hunk 3 above)
Comment (Duyi-Wang) [reply modelzoo/DIN/train.py:512]: I just use `args.steps`  when I want to test the model, get the model's timeline, and benchmark gsteps. When tuning ACC/AUC, I can just use the command `python train.py` for training a specific number of epochs.

File: modelzoo/DIN/train.py (hunk 2 above)
Comment (aalbersk) [reply modelzoo/DIN/train.py:413]: That's interesting. Could you point me to part of the code where I can see these optimizations?

File: modelzoo/DIN/train.py (hunk 3 above)
Comment (aalbersk) [reply modelzoo/DIN/train.py:512]: Fair enough

File: modelzoo/DIN/train.py (hunk 2 above)
Comment (Duyi-Wang) [reply modelzoo/DIN/train.py:413]: You can see the recently [Optimizer] commits in Alibaba/DeepRec. DeepRec does some fusion on this op.
